pypicloud/cache/dynamo.py

```diff
@@ -22,6 +25,78 @@
 LOG = logging.getLogger(__name__)
 
 
+def _path_depth(package):
+    """Count the slashes in path, or return infinity on no path."""
+    path = package.data.get('path')
+    if not path:
+        return float('+inf')
+    return path.count('/')
+
+
+def _decide_between_versions(contender, current):
+    """Decide between packages with the same filename on different paths."""
+
+    if contender.last_modified < current.last_modified:
+        return contender
```

Comment:
Sorry, I don't mean to waste your time, but I am either misunderstanding some of your points or misreading your code. I'm going to try to nail down the parts I find confusing.

> if they have the same path, the storage always wins.

I don't think this is the case. That's what [the docstring on _decide_between_versions](https://github.com/mathcamp/pypicloud/pull/69/files#diff-cd8bb69a797363ac5abd920593f7114bR37) says, but it looks like it's being called on [every pair of packages](https://github.com/mathcamp/pypicloud/pull/69/files#diff-cd8bb69a797363ac5abd920593f7114bR86) where the same filename appears in both the cache and the storage, regardless of if the paths differ.

If the paths *do* differ, we want to keep the one that's currently in the cache to avoid nasty surprises. But in `_decide_between_versions` the priority is first decided by `last_modified` and then `path_depth`. Shouldn't the first operation be to compare paths and, if unequal, return `current`? (Preferably with a log warning that says that something fishy is going on)

> Considering the cache more than just a cache, since the values that it stores get decided based on both the storage and itself, is something that asks for trouble

Isn't the rebuild operation doing exactly that? It's comparing the versions of packages in the cache & storage before determining which updates get pushed to the cache.